### ISSUE TITLE
[=BUG=] ReadError: The server aborted pending request

### ISSUE THREAD

**Issue by cjh980402** [play-dl/play-dl#66]
#### Describe the bug

 - Node.js version: 16.9.0
 - OS & version: ubuntu 20.04

I play this song (https://www.youtube.com/watch?v=NGYYKJ0R0OA) at 21:20. And I got this error (`ReadError: The server aborted pending request`) at 21:32.

```js
url: https://www.youtube.com/watch?v=NGYYKJ0R0OA
duration: 8226
Error stack: ReadError: The server aborted pending request
    at IncomingMessage.<anonymous> (/home/ubuntu/bot/node_modules/got/dist/source/core/index.js:803:31)
    at Object.onceWrapper (node:events:513:28)
    at IncomingMessage.emit (node:events:406:35)
    at IncomingMessage.origin.emit (/home/ubuntu/bot/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)
    at IncomingMessage._destroy (node:_http_incoming:179:10)
    at _destroy (node:internal/streams/destroy:102:25)
    at IncomingMessage.destroy (node:internal/streams/destroy:64:5)
    at TLSSocket.socketCloseListener (node:_http_client:407:11)
    at TLSSocket.emit (node:events:406:35)
    at node:net:672:12
```

#### Actual behavior

Sometimes, I got the ReadError.

#### Expected behavior

 
It should be no error.

#### Code to reproduce

```js

```

#### Checklist

- [x] I have read the documentation.
- [x] I have tried my code with the latest version of Node.js and Play-dl.

**Comment by killer069** [comment]
@cjh980402 

Told you old loop was quite efficient. Now enjoy until I change loop back to normal.

**Comment by cjh980402** [comment]
I think it was my fault too :(

**Comment by cjh980402** [comment]
New loop has aborted error like this issue.

And old loop is runned every 280 seconds. Then, in old loop, even if we pause the song, it is runned every 280 seconds. After some times, even if stream is fulled (highWaternark value), old loop is runned every 280 seconds. Then, what happened in old loop?

**Comment by killer069** [comment]
> New loop has aborted error like this issue.
> 
> And old loop is runned every 280 seconds. Then, in old loop, even if we pause the song, it is runned every 280 seconds. After some times, even if stream is fulled (highWaternark value), old loop is runned every 280 seconds. Then, what happened in old loop?

You can't handle that in new loop either. So no need to talk about that.

But with 280 seconds loop, at least YouTube doesn't rate limit us like this issue.

**Comment by killer069** [comment]
@cjh980402 

I didn't touched your end concept. Just added a timer to it.

That end and content-length was cool.

**Comment by cjh980402** [comment]
Thanks :)

**Comment by cjh980402** [comment]
@killer069 I have some idea(question). How about this?
I think we should check 'end' event to prevent over-piped and some problem about pause the song long time, but run loop immediately in 'end' event we got the aborted error.
Then how about 'end' event + timer?

```js
private loop(){
    if(this.stream.destroyed){
        this.cleanup()
        return
    }
    let after280 = Date.now() + 280 * 1000;
    let end : number = this.bytes_count + this.per_sec_bytes * 300;
    let stream = got.stream(this.url, {
        headers : {
            "range" : `bytes=${this.bytes_count}-${end >= this.content_length ? '' : end}`
        }
    })
    this.request = stream
    stream.pipe(this.stream, { end : false })

    stream.once('error', (err) => {
        this.stream.emit('error', err)
    })

    stream.on('data', (chunk: any) => {
        this.bytes_count += chunk.length
    })

    stream.on('end', () => {
        if(end < this.content_length) {
            setTimeout(() => this.loop(), Math.max(0, after280 - Date.now()))
        }
        else this.cleanup()
    })
}
```

**Comment by killer069** [comment]
> @killer069 I have some idea(question). How about this?
> I think we should check 'end' event to prevent over-piped and some problem about pause the song long time, but run loop immediately in 'end' event we got the aborted error.
> Then how about 'end' event + timer?
> 
> ```js
> private loop(){
>     if(this.stream.destroyed){
>         this.cleanup()
>         return
>     }
>     let after280 = Date.now() + 280 * 1000;
>     let end : number = this.bytes_count + this.per_sec_bytes * 300;
>     let stream = got.stream(this.url, {
>         headers : {
>             "range" : `bytes=${this.bytes_count}-${end >= this.content_length ? '' : end}`
>         }
>     })
>     this.request = stream
>     stream.pipe(this.stream, { end : false })
> 
>     stream.once('error', (err) => {
>         this.stream.emit('error', err)
>     })
> 
>     stream.on('data', (chunk: any) => {
>         this.bytes_count += chunk.length
>     })
> 
>     stream.on('end', () => {
>         if(end < this.content_length) {
>             setTimeout(() => this.loop(), Math.max(0, after280 - Date.now()))
>         }
>         else this.cleanup()
>     })
> }
> ```

I have a better approach, do some testing on developer branch. I have fixed all issues.

**Comment by cjh980402** [comment]
Oh, I see the developer branch after write this haha

**Comment by killer069** [comment]
@cjh980402 

I am doing some testing and then will push in main branch.

**Comment by killer069** [comment]
@cjh980402 

Now test and tell me about new loop.

**Comment by cjh980402** [comment]
It works good in my test (no pause error and other error)

**Comment by killer069** [comment]
@cjh980402 

Can you run some tests on `play-dl-https` at [here](https://github.com/play-dl/play-dl-https) ?
Also compare the speed of both.

If you do, Thanks in advance.

My plan is to make play-dl to 1 dependency package [ @types/node only ].

**Comment by cjh980402** [comment]
Oh, it doesn't use got module! I'll test it soon.

**Comment by killer069** [comment]
Thanks

**Comment by cjh980402** [comment]
In my speed test of stream,
1st url: https://www.youtube.com/watch?v=rn0wEJmoGdg
play-dl: 1396ms
play-dl-https: 1749ms

2nd url: https://www.youtube.com/watch?v=wp7l9oolXS0
play-dl: 1560ms
play-dl-https: 1785ms

Almost same both. Also, `play-dl-https` didn't make any error in my test. (I didn't check all the cases)

**Comment by killer069** [comment]
> In my speed test of stream,
> 1st url: https://www.youtube.com/watch?v=rn0wEJmoGdg
> play-dl: 1396ms
> play-dl-https: 1749ms
> 
> 2nd url: https://www.youtube.com/watch?v=wp7l9oolXS0
> play-dl: 1560ms
> play-dl-https: 1785ms
> 
> Almost same both. Also, `play-dl-https` didn't make any error when play the song in my test. (I didn't check all the cases)

Thanks for testing. So it means it is quite stable and creates stream quite fast.
got results are kinda fake as it tends to return stream first and plays the song with a delay.

Adding that delay makes got and https almost same. So in next update, I would be shifting to https soon.

**Comment by cjh980402** [comment]
@killer069 
Sometimes I got this error(`ReadError: The server aborted pending request`) again when play the very long song (longer than 3hours).

1st
url: https://www.youtube.com/watch?v=AITuNQSCR6k
duration: 3:20:42
error time: about 3:10:00

2nd
url: https://www.youtube.com/watch?v=FQtU0pqHcpk
duration: 3:40:08
error time: about 2:44:00

3rd
url: https://www.youtube.com/watch?v=macF3chXBgM
duration: 6:18:45
error time: about 4:00:00

In my case, I didn't get this error when song is shorter than 2.5hours.

**Comment by killer069** [comment]
@cjh980402 

Thanks for checking long duration playback. I will update playing url every 2 hours [ just like I update url for Live Stream every 30 mins ]. This will be included in next update.

**Comment by Eralyne** [comment]
I see the issue is closed. Is this solved in 1.0.1 or going to be looked into in a future update? @killer069

**Comment by killer069** [comment]
> I see the issue is closed. Is this solved in 1.0.1 or going to be looked into in a future update? @killer069

It has been patched in 1.0.1.
